### PR TITLE
Pytorch 0.2

### PR DESCRIPTION
**Pull request by DeNeutoy**

### CODE REVIEW

File: Dockerfile.cpu [hunk 1]
```diff
@@ -44,7 +44,7 @@ COPY scripts/ scripts/
 RUN find . | grep -E "(__pycache__|\.pyc|\.pyo$)" | xargs rm -rf
 
 RUN /bin/bash -c 'source activate allennlp && INSTALL_TEST_REQUIREMENTS="true" ./scripts/install_requirements.sh'
-RUN /bin/bash -c 'source activate allennlp && pip install --no-cache-dir -q http://download.pytorch.org/whl/cu75/torch-0.1.12.post2-cp35-cp35m-linux_x86_64.whl'
+RUN /bin/bash -c 'source activate allennlp && pip install --no-cache-dir -q http://download.pytorch.org/whl/cu80/torch-0.2.0.post1-cp35-cp35m-manylinux1_x86_64.whl'
```
Comment (matt-gardner) [inline Dockerfile.cpu]: @joelgrus, @schmmd - these two files look identical, why do we have two dockerfiles?

File: allennlp/models/decomposable_attention.py [hunk 2]
```diff
@@ -143,24 +143,22 @@ def forward(self,  # type: ignore
         hypothesis_compare_input = torch.cat([embedded_hypothesis, attended_premise], dim=-1)
 
         compared_premise = self._compare_feedforward(premise_compare_input)
-        # TODO(mattg): use broadcasting once pytorch 0.2 is released.
-        compared_premise = compared_premise * premise_mask.unsqueeze(-1).expand_as(compared_premise)
+        compared_premise = compared_premise * premise_mask.unsqueeze(-1)
```
Comment (matt-gardner) [inline allennlp/models/decomposable_attention.py]: Do we need the unsqueeze here?

File: allennlp/models/decomposable_attention.py [hunk 3]
```diff
@@ -143,24 +143,22 @@ def forward(self,  # type: ignore
...
         # Shape: (batch_size, compare_dim)
-        compared_premise = compared_premise.sum(dim=1).squeeze(1)
+        compared_premise = compared_premise.sum(dim=1)
 
         compared_hypothesis = self._compare_feedforward(hypothesis_compare_input)
-        # TODO(mattg): use broadcasting once pytorch 0.2 is released.
-        compared_hypothesis = compared_hypothesis * hypothesis_mask.unsqueeze(-1).expand_as(compared_hypothesis)
+        compared_hypothesis = compared_hypothesis * hypothesis_mask.unsqueeze(-1)
```
Comment (matt-gardner) [inline allennlp/models/decomposable_attention.py]: Also here.

File: allennlp/modules/seq2vec_encoders/cnn_encoder.py [hunk 4]
```diff
@@ -101,7 +101,7 @@ def forward(self, tokens: torch.Tensor, **kwargs):  # pylint: disable=arguments-
         # projected using the projection layer, if requested.
 
         # Not really sure why max isn't squeezing...
```
Comment (matt-gardner) [inline allennlp/modules/seq2vec_encoders/cnn_encoder.py]: Remove this comment.

File: allennlp/modules/token_embedders/embedding.py [hunk 5]
```diff
@@ -104,15 +105,16 @@ def get_output_dim(self) -> int:
...
         original_inputs = inputs
         if original_inputs.dim() > 2:
             inputs = inputs.view(-1, inputs.size(-1))
-        embedded = self._backend.Embedding(padding_index,
-                                           self.max_norm,
-                                           self.norm_type,
-                                           self.scale_grad_by_freq,
-                                           self.sparse)(inputs, self.weight)
+        embedded = embedding(inputs, self.weight,
+
```
Comment (matt-gardner) [inline allennlp/modules/token_embedders/embedding.py]: Extra line here.

File: allennlp/nn/util.py [hunk 6]
```diff
@@ -126,13 +126,11 @@ def arrays_to_variables(data_structure: Dict[str, Union[dict, numpy.ndarray]],
 def _get_normalized_masked_log_probablities(vector, mask):
     # We calculate normalized log probabilities in a numerically stable fashion, as done
     # in https://github.com/rkadlec/asreader/blob/master/asreader/custombricks/softmax_mask_bricks.py
-    # TODO(mattg): a bunch of this logic can be simplified once pytorch-0.2 is out.
-    # torch.max(keepdim=True), for instance, simplifies things here.
     input_masked = mask * vector
-    shifted = mask * (input_masked - torch.max(input_masked, dim=1)[0].expand_as(input_masked))
+    shifted = mask * (input_masked - input_masked.max(dim=1, keepdim=True)[0])
```
Comment (matt-gardner) [inline allennlp/nn/util.py]: Hmm, with broadcasting, do we even need the `keepdim=True`, here and below?

File: allennlp/nn/util.py [hunk 7]
```diff
@@ -344,23 +342,24 @@ def sequence_cross_entropy_with_logits(logits: torch.FloatTensor,
...
-    # back in the `replace_with` value.
-    # TODO(mattg): use broadcasting here when it's available in pytorch 0.2.
-    if tensor.dim() != mask.dim():
-        raise ConfigurationError("tensor.dim() (%d) != mask.dim() (%d)" % (tensor.dim(), mask.dim()))
-    one_minus_mask = 1.0 - mask.expand_as(tensor)
+    # We'll build a tensor of the same shape as `tensor`, subtract away
+    # masked values, then add back in the `replace_with` value.
+    if tensor.dim() < mask.dim():
+        raise ConfigurationError("Mask has dimension {} but tensor has dimension {}. Mask dimension "
+                                 "should be <= tensor dimension.".format(mask.dim(), tensor.dim()))
```
Comment (matt-gardner) [inline allennlp/nn/util.py]: I'm not sure this `.format()` call works as you expect.  I _think_ it'll get called on the second string before the implicit concatenation.  Any idea if that's true or not?

File: allennlp/nn/util.py [hunk 8]
```diff
@@ -344,23 +342,24 @@ def sequence_cross_entropy_with_logits(logits: torch.FloatTensor,
...
+    # Expand mask dimensions to match tensor if it's smaller.
+    while mask.dim() < tensor.dim():
+        mask = mask.unsqueeze(-1)
```
Comment (matt-gardner) [inline allennlp/nn/util.py]: I'm actually not sure we can do this safely.  I think my original comment about requiring `tensor.dim() == mask.dim()` only until pytorch 0.2 was wrong, and we probably should just leave this as it was (except we can probably remove the `expand_as`).

Consider:

If this is for an embedding, we have tensor with shape `(batch_size, num_tokens, embedding_dim)`, with mask `(batch_size, num_tokens)`.  We want to unsqueeze the last dimension, and that's fine.  But in BiDAF, I use this with a tensor of shape `(batch_size, passage_length, question_length)`, with a mask of shape `(batch_size, question_length)`.  Unsqueezing the last dimension here is wrong, and could lead to odd and unpredictable behavior of `passage_length` happens to equal `question_length` - the mask will be applied on the wrong dimension.  I think we just need to require the caller to have put the mask into the same dimension as the tensor, so we don't have to guess in this function.

I'm a little concerned that this passes the BiDAF test, actually - I'm not sure how this passes.

File: Dockerfile.cpu (hunk 1 above)
Comment (joelgrus) [inline Dockerfile.cpu]: the gpu one uses the cuda base image, the cpu one uses a plain ubuntu base image. (previously they were pointing at different pip wheels, as per the pytorch documentation, but that might not be true in pytorch 2.0)

Comment (matt-gardner) [inline Dockerfile.cpu]: Thanks, sorry for the dumb question.  I typed the wrong command in my terminal when trying to diff the two dockerfiles, and so it looked like there was no difference bewteen them...

File: allennlp/models/decomposable_attention.py (hunk 2 above)
Comment (DeNeutoy) [inline allennlp/models/decomposable_attention.py]: It seems so, unfortunately. 

```
x = torch.rand([3,4,5])
y = torch.rand([4,5])
x * y # fine.

y = torch.rand([3,4]) # most common mask broadcasting case.
x * y # raises.
```

File: Dockerfile.cpu (hunk 1 above)
Comment (DeNeutoy) [inline Dockerfile.cpu]: The different pip wheel seems to just change the supported CUDA package (for None they just install the cu75 wheel), so don't think it makes any difference. I don't see why having CUDA 7.5 would be lighter than CUDA 8, so I guess we can stick with this unless it actually breaks anything?

File: allennlp/nn/util.py (hunk 7 above)
Comment (DeNeutoy) [inline allennlp/nn/util.py]: Nope, looks like it works fine (it works split in a python interpreter at least).

File: allennlp/nn/util.py (hunk 8 above)
Comment (DeNeutoy) [inline allennlp/nn/util.py]: Yeah ... I wasn't quite sure when doing this that this was what you meant - i'll reverse it. The bidaf tests are still passing because I didn't remove the `mask.unsqueeze(-1)` before the function is called, making the mask the same shape as the tensor anyway.

File: allennlp/nn/util.py (hunk 7 above)
Comment (matt-gardner) [inline allennlp/nn/util.py]: Huh, that's good to know.  Here I've been avoiding things like this thinking it wouldn't work...

File: allennlp/modules/token_embedders/embedding.py (hunk 5 above)
Comment (matt-gardner) [inline allennlp/modules/token_embedders/embedding.py]: Looks like you missed removing this blank line.
